### PR TITLE
server: add ForceServerCodec() to set a custom encoding.Codec on the server

### PR DESCRIPTION
**Pull request by ash2k**
Fixes https://github.com/grpc/grpc-go/issues/3694.

### CODE REVIEW

**Comment by ash2k** [comment]
I'm sure the godoc needs tweaking, but just wanted to start somewhere.

File: rpc_util.go [hunk 1]
```diff
@@ -429,9 +429,9 @@ func (o ContentSubtypeCallOption) before(c *callInfo) error {
 }
 func (o ContentSubtypeCallOption) after(c *callInfo, attempt *csAttempt) {}
 
-// ForceCodec returns a CallOption that will set the given Codec to be
+// ForceCodec returns a CallOption that will set the given codec to be
```
Comment (ash2k) [inline rpc_util.go:432]: IntelliJ recognizes `Codec` as a reference to the deprecated type, which it is not. So I thought it's ok to change it to the reference to the formal parameter instead.

File: server.go [hunk 2]
```diff
@@ -266,6 +266,29 @@ func CustomCodec(codec Codec) ServerOption {
...
+// content-subtype.
+//
+// This function is provided for advanced users; prefer to use only
+// CallContentSubtype to select a registered codec on the client instead.
+//
+// Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+func ForcedCodec(codec encoding.Codec) ServerOption {
```
Comment (ash2k) [inline server.go:286]: I don't know how to call it, please suggest.

Comment (dfawley) [reply server.go:286]: Yes, the lack of namespacing here is an issue.

Let's try `ForceServerCodec()`.

File: server.go [hunk 3]
```diff
@@ -266,6 +266,29 @@ func CustomCodec(codec Codec) ServerOption {
...
+// This will override any lookups by content-subtype for Codecs registered with RegisterCodec.
+//
+// See Content-Type on
+// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests for
+// more details. Also see the documentation on RegisterCodec and
+// CallContentSubtype for more details on the interaction between encoding.Codec and
+// content-subtype.
+//
+// This function is provided for advanced users; prefer to use only
+// CallContentSubtype to select a registered codec on the client instead.
```
Comment (dfawley) [inline server.go:280]: (Mostly from `CustomCodec`):

...; prefer to register codecs using encoding.RegisterCodec. The server will automatically use registered codecs based on the incoming requests' headers. See also https://github.com/grpc/grpc-go/blob/master/Documentation/encoding.md#using-a-codec. Will be supported throughout 1.x.

File: server.go [hunk 4]
```diff
@@ -266,6 +266,29 @@ func CustomCodec(codec Codec) ServerOption {
...
+//
+// Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+func ForcedCodec(codec encoding.Codec) ServerOption {
+	return newFuncServerOption(func(o *serverOptions) {
+		o.codec = codec
```
Comment (dfawley) [inline server.go:288]: Well this was surprisingly simple.  :)

Could you please add tests in `test/end2end_test.go` for this?  See https://github.com/grpc/grpc-go/blob/4f3aa7cfa157c38bd5c2da7f4568614f815ab4ad/test/end2end_test.go#L5260 for an example of the style of test we use now.  (`Start` accepts `ServerOption`s already, so it should be pretty simple to get working.)

**Review by dfawley — CHANGES_REQUESTED** [review]

**Comment by dfawley** [comment]
Oh, please also rebase your branch based on master@HEAD; we have new tests and some flaky test fixes that will be simpler to deal with if you do.  Thanks!

**Comment by ash2k** [comment]
@dfawley Thanks for the review, I've addressed your comments. Please let me know if something is not right.

File: server.go [hunk 5]
```diff
@@ -279,6 +279,31 @@ func CustomCodec(codec Codec) ServerOption {
 	})
 }
 
+// ForceServerCodec returns a ServerOption that sets a codec for message marshaling and unmarshaling.
```
Comment (dfawley) [inline server.go:282]: Please wrap comment blocks to 80 columns.

**Review by dfawley — COMMENTED** [review]
The new test looks great, thanks!  Just a few minor style things inline.

**Comment by ash2k** [comment]
@dfawley Done.

**Review by dfawley — APPROVED** [review]
Thank you for the PR, and sorry about the slow turnaround time!

**Pull request merged**